SmartDeviceLink/private/SDLPresentAlertOperation.m

```diff
@@ -0,0 +1,463 @@
+}
+
+#pragma mark - Cancel
+
+/// Cancels the alert. If the alert has not yet been sent to the module, it will not be sent. If the alert is already presented on the module, the alert will be immediately dismissed. Canceling an already presented alert will only work if connected to modules supporting RPC spec versions 6.0+. On older versions alert will not be dismissed.
+- (void)sdl_cancelInteraction {
+    if (self.isFinished) {
+        SDLLogW(@"This operation has already finished so it can not be canceled.");
+        return;
+    } else if (self.isCancelled) {
+        SDLLogW(@"This operation has already been canceled. It will be finished at some point during the operation.");
+        return;
+    } else if (self.isExecuting) {
+        if ([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
+            SDLLogE(@"Canceling an alert is not supported on this module");
+            [self cancel];
+            return;
+        }
+
+        SDLLogD(@"Canceling the presented alert");
+        SDLCancelInteraction *cancelInteraction = [[SDLCancelInteraction alloc] initWithAlertCancelID:self.cancelId];
+
+        __weak typeof(self) weakSelf = self;
+        [self.connectionManager sendConnectionRequest:cancelInteraction withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+            if (error != nil) {
+                weakSelf.internalError = error;
+                SDLLogE(@"Error canceling the presented alert: %@, with error: %@", request, error);
+                return;
+            }
+            SDLLogD(@"The presented alert was canceled successfully");
+        }];
+    } else {
+        SDLLogD(@"Canceling an alert that has not yet been sent to Core");
+        [self cancel];
+    }
+}
+
+#pragma mark - Private Getters / Setters
+
+/// Assembles an `Alert` RPC from the `SDLAlertView` information.
+/// @return The `Alert` RPC to be sent to the module.
+- (SDLAlert *)alertRPC {
+    SDLAlert *alert = [[SDLAlert alloc] init];
+    [self sdl_assembleAlertText:alert];
+    alert.duration = @((NSUInteger)(self.alertView.timeout * 1000));
+    alert.alertIcon = [self sdl_supportsAlertIcon] ? self.alertView.icon.imageRPC : nil;
+    alert.progressIndicator = @(self.alertView.showWaitIndicator);
+    alert.cancelID = @(self.cancelId);
+
+    // The number of alert soft buttons sent must be capped so there are no clashes with soft button ids assigned by other managers (And thus leading to clashes saving/retreiving the button handlers in the  `SDLResponseDispatcher` class)
+    NSUInteger softButtonCount = MIN(self.alertView.softButtons.count, SDLAlertSoftButtonCount);
+    NSMutableArray<SDLSoftButton *> *softButtons = [NSMutableArray arrayWithCapacity:softButtonCount];
+    for (NSUInteger i = 0; i < softButtonCount; i += 1) {
+        SDLSoftButtonObject *button = self.alertView.softButtons[i];
```

Comment:
Since there can be issues with omitting the image for an "image-only" soft button (i.e. do we omit just that soft button or do we not send any of the soft buttons) we decided to mirror the behavior of the soft button manager and just send the soft buttons as is even if the image upload fails.